build_steps/Dockerfile

```diff
@@ -0,0 +1,24 @@
+FROM nvidia/cuda:11.5.0-devel-ubuntu20.04
+
+ARG USERNAME="pygennuser"
+ARG PYGENN_HOME="/root"
+
+RUN apt-get update && \
+    apt-get upgrade -y
+
+RUN apt-get install -yq --no-install-recommends curl vim python3-dev python3-pip
+
+RUN pip install swig numpy jupyter
+
+ENV CUDA_PATH=/usr/local/cuda-11.5
+ENV HOME=${PYGENN_HOME}
+ENV USERNAME=${USERNAME}
+
+WORKDIR ${HOME}
+
+COPY . ${HOME}
+
+RUN adduser --disabled-password --gecos "" $USERNAME
```

Comment:
Why not just use root as containers are kinda throwaway? My knowledge of docker is pretty crap so I could very well be missing something...

Reply:
It’s something I was taught was best practice so have got into the habit of doing. I believe this is because:
1. Certain execution environments block containers running as root
2. Force container immutability
3. Limits the processes that can be executed and who can execute them

Reply:
Ahh, that is really good to know, thanks

Reply:
Running the container as root (from the host) could be bad, sure, but I can't see any issue being root *inside* the container. The mutability argument, I'm not sure how relevant that is, since all state is lost once the container is killed anyway. As for limiting processes, sure, it stops the user running e.g. `apt` inside the container as interactive, but is that a good thing?

Reply:
Btw, I don't mean to be pedantic. I'm no expert, and I'm very interested to hear yours and others' reasons for and against!

Reply:
Ahh.... I see the issue with containers starting as root now..... 

I just pwnt by box with a container that begins as root! :) 

https://flast101.github.io/docker-privesc/